### PR TITLE
Rust generator

### PR DESCRIPTION
**Pull request by boxdot**
Also extend the coappearances examples to show how to use optional subarchives, which do not have to be serialized directly and may be added later.

### CODE REVIEW

**Comment by boxdot** [comment]
@imagovrn: I split up the Rust generator templates in files and added tests. I also removed `is_index` method from `Struct` node. Instead, I added a backreference to `multivector` which references the corresponding index type.

Two features are still missing:
* enum types
* optional resources

I would like to add those to the rust generator after I implemented them in https://github.com/boxdot/flatdata-rs.

File: generator/generators/BaseGenerator.py [hunk 1]
```diff
@@ -32,6 +32,8 @@ def render(self, tree):
         env.filters['is_raw_data'] = lambda n: isinstance(n, RawData)
         env.filters['is_archive_resource'] = lambda n: isinstance(n, ArchiveResource)
         env.filters['is_structure'] = lambda n: isinstance(n, Structure)
+        env.filters['is_index'] = lambda n: (isinstance(n, Structure) and
```
Comment (VeaaC) [inline generator/generators/BaseGenerator.py]: is_builtin_structure? We might get more than the simple integer based ones

File: generator/generators/RustGenerator.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    Archive as ArchiveResource)
+from generator.tree.nodes.trivial import Structure, Constant
+from generator.tree.nodes.archive import Archive
+from .BaseGenerator import BaseGenerator
+
+import re
+
+class RustGenerator(BaseGenerator):
+
+    RESERVED_KEYWORDS = [
```
Comment (VeaaC) [inline generator/generators/RustGenerator.py]: Nice, we might want to add this for other languages as well :-)

File: generator/generators/RustGenerator.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        return [Structure, Archive, Constant]
+
+    def _populate_environment(self, env):
+        def _camel_to_snake_case(s):
+            s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', s)
+            return re.sub('([a-z0-9])(A-Z)', r'\1_\2', s1).lower()
+
+        env.filters["camel_to_snake_case"] = _camel_to_snake_case
+
+        def _snake_to_camel_case(s):
```
Comment (VeaaC) [inline generator/generators/RustGenerator.py]: There are two type of camel case, starting with upper case or not. Maybe be more specific here?

File: generator/generators/RustGenerator.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        env.filters["rust_doc"] = _rust_doc
+
+        def _escape_rust_keywords(s):
+            if s in self.RESERVED_KEYWORDS:
+                return "{}_".format(s)
+            return s
+
+        env.filters["escape_rust_keywords"] = _escape_rust_keywords
+
+        env.filters['instance_resources'] = lambda ls: [
```
Comment (VeaaC) [inline generator/generators/RustGenerator.py]: Thought: We could add those to the node then they would available in all generators?

File: generator/templates/rust/rust.jinja2 [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{% import "rust/constant.jinja2" as constant %}
+{# {% import "cpp/enumeration.jinja2" as enumeration %} #}
+{% import "rust/structure.jinja2" as structure %}
+{% import "rust/archive.jinja2" as archive %}
+// Do not edit: This code was generated by flatdata's generator.
+
+use flatdata::Struct;
+
+{% macro render_namespaces(root) %}
+{% for module in root.children if module | is_namespace %}
```
Comment (VeaaC) [inline generator/templates/rust/rust.jinja2]: What if we do not have any namespace?

File: generator/generators/BaseGenerator.py (hunk 1 above)
Comment (boxdot) [reply generator/generators/BaseGenerator.py]: In the next line I actually check that this an index.

File: generator/generators/RustGenerator.py (hunk 2 above)
Comment (boxdot) [reply generator/generators/RustGenerator.py]: It was quite easy to find a list of keywords for Rust. Also not that hard to find them for C++: http://en.cppreference.com/w/cpp/keyword. I will do same for C++.

File: generator/generators/RustGenerator.py (hunk 3 above)
Comment (boxdot) [reply generator/generators/RustGenerator.py]: Renamed to `snake_to_upper_camel_case`.

File: generator/generators/RustGenerator.py (hunk 4 above)
Comment (boxdot) [reply generator/generators/RustGenerator.py]: Do you mean "to add them to `BaseGenerator`"?

File: generator/templates/rust/rust.jinja2 (hunk 5 above)
Comment (boxdot) [reply generator/templates/rust/rust.jinja2]: Not having a root namespace is an invalid schema. I checked.

File: generator/generators/BaseGenerator.py (hunk 1 above)
Comment (VeaaC) [reply generator/generators/BaseGenerator.py]: maybe then name it multi_vector_index?

File: generator/generators/RustGenerator.py (hunk 4 above)
Comment (VeaaC) [reply generator/generators/RustGenerator.py]: I meant adding them to the syntax tree's node (similar to other node types which can list certain types of children)

**Comment by boxdot** [comment]
All features are implemented except for nested namespaces. I don't see right now an easy way to represent them in Rust. So, I would rather prefer just to flatten them in a single global one.

File: generator/generators/BaseGenerator.py (hunk 1 above)
Comment (boxdot) [reply generator/generators/BaseGenerator.py]: Fixed

File: generator/tree/nodes/resources/multivector.py [hunk 6]
```diff
@@ -24,13 +25,25 @@ def create(properties, own_schema):
     def _create_references(self):
         return [StructureReference(name=t) for t in self._types]
 
+    @property
```
Comment (boxdot) [inline generator/tree/nodes/resources/multivector.py]: This is a fix. Was missing.

File: generator/generators/BaseGenerator.py [hunk 7]
```diff
@@ -39,6 +39,8 @@ def render(self, tree):
         env.filters['is_bound_resource'] = lambda n: isinstance(n, BoundResource)
         env.filters['is_vector'] = lambda n: isinstance(n, Vector)
         env.filters['is_multivector'] = lambda n: isinstance(n, Multivector)
+        env.filters['is_multivector_index'] = lambda n: (isinstance(n, Structure) and
+            "_builtin.multivector" in SyntaxTree.namespace_path(n))
```
Comment (imagovrn) [inline generator/generators/BaseGenerator.py]: `startswith` instead of `in`?

File: generator/generators/RustGenerator.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            lines = [
+                re.sub(r'^[ \t]*(/\*\*|/\*|\*/|\*)\s*(.*?)\s*(\*/)?$', r"/// \2", line).strip()
+                for line in s.split('\n')
+            ]
+            start = 0
+            end = len(lines)
+            if lines[0] == "///":
+                start = 1
+            if lines[-1] == "///":
+                end = -1;
```
Comment (imagovrn) [inline generator/generators/RustGenerator.py]: I am not sure what intended here, but maybe in a same fashion as previous `if` clause it should be `-2`?

File: generator/generators/RustGenerator.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            return "\n".join(lines[start:end])
```
Comment (imagovrn) [inline generator/generators/RustGenerator.py]: `_rust_doc` might not work well on empty `lines`

**Pull request merged**